### DISCUSSION TITLE
ストリーミングレプリケーションが上手く行かない？

### DISCUSSION THREAD

**Discussion by askdkc** [pgroonga/pgroonga#289] [Q&A] [answered]
お世話になります

https://pgroonga.github.io/ja/reference/streaming-replication.html

↑ こちらの手順を実行してUbuntu 22.04環境（`postgresql-14/jammy-updates,now 14.6-0ubuntu0.22.04.1 amd64`、`postgresql-14-pgroonga/jammy,now 2.4.4-1.ubuntu22.04.1 amd64`）でPGroongaのストリーミングレプリケーション環境を試験的に作ったのですが、下記のように`pg_basebackup`以後のデータ反映に問題が出ている状況です

- スタンバイ側で`pg_basebackup`を実行後、元々プライマリで作成されたデータのPGroongaのインデックス検索は問題なく動く
- プライマリ側で`createdb newdb`をし、データを追加した後、スタンバイ側で`LIKE`検索等は問題なく出来る
- プライマリ側で`newdb`に`CREATE EXTENSION pgroonga`をし、PGroongaのインデックスを作成後にスタンバイ側で検索をすると`ERROR:  pgroonga: object isn't found: <Sources19169>`のようなエラーが出力される
- `pg_basebackup`実行時に存在していたプライマリ側のDBに新規にデータを追加してもスタンバイ側に`select * from table_name`では確認可能だが、`LIKE`検索や`&@~`検索ではヒットしない(インデックスがいない？)

というような感じで

https://pgroonga.github.io/ja/reference/streaming-replication.html

の「pg_basebacup以降にプライマリーで追加したデータも検索できます。」が上手く行かない状況です。

どの辺を調べれば良いのかお分かりになりますでしょうか？

**Comment by komainu8** [comment]
ご報告ありがとうございます。
手元でも再現するか確認中なので少々お待ちください。

**Reply by askdkc** [reply]
ありがとうございます

最初試していた環境を確認したところ1個だけ上手くDBがあったりしたのですが（プライマリにデータを追加すると、セカンダリのPGroonga検索でも反映される期待通りの動作）、初めから構築をやり直してみると上手く行かなかったり、createdbでプライマリに新たに追加したDBはセカンダリでも相変わらず`ERROR:  pgroonga: object isn't found: <Sources19169>`みたいなエラーが出たりと動作が現状謎です

余裕があればUbuntu 20.04の環境とかでも試してみたいと思います

**Comment by askdkc** [comment]
追加で試したところ

https://pgroonga.github.io/ja/reference/streaming-replication.html

の手順に従って1つだけDBを作理、そこにPGroongaインデックスを張ったDBは上手く動くようです

ただ、最初から複数のDBを作成し、それぞれにPGroongaのインデックスを持たせた後にセカンダリでpg_basebackupをしても、pg_basebackup時のインデックスは検索可能なのですが、それ以後プライマリで新規に追加したデータについては一切インデックスデータがセカンダリに反映されないようでした

**Comment by askdkc** [comment, marked as answer]
多分原因が分かったのですが、チュートリアルを愚直にやっていところセカンダリ側の`postgresql.conf`に

```vim
pgroonga.enable_wal = on
pgroonga.max_wal_size = 100MB
```

の記載が漏れていました

こちらを設定してPostgreSQLを再起動したところ、新規でプライマリにどれだけDBを作成しても、何もかも、全て上手く動くようになりました

お騒がせしました〜🙇‍♂️

**Reply by kou** [reply]
あれ、スタンバイで`pg_basebackup`するとプライマリーのその設定もコピーされませんでしたっけ？

**Reply by askdkc** [reply]
@kou さん

Ubuntu 22.04/20.04で試してたのですが、なんか上手く行かなかってですね〜
謎です〜

**Reply by kou** [reply]
PostgreSQLのバージョンアップで挙動が変わったんですかねぇ。
変わっているならドキュメントを更新しないといけないですねぇ。

**Reply by askdkc** [reply]
@kou さん

振りですね。
わかります。

**Reply by askdkc** [reply]
PR出しておきました〜(キリ番100だわ💓)

https://github.com/pgroonga/pgroonga.github.io/pull/100
